src/qiboml/models/decoding.py

```diff
@@ -283,3 +285,47 @@ def output_shape(self) -> tuple[int, int]:
     @property
     def analytic(self) -> bool:  # pragma: no cover
         return False
+    
+@dataclass
+class VariationalQuantumLinearSolver(QuantumDecoding):
+    """Decoder for the Variational Quantum Linear Solver (VQLS).
+    Adapted from the following paper arXiv:1909.05820v4 by Carlos Bravo-Prieto et al. 
+
+    Args:
+        target_state (ndarray): Target solution vector |b⟩.
+        A (ndarray): The matrix A in the linear system A|x⟩ = |b⟩.
+    
+    Ensure QiboML backend set to Pytorch. 
+    """
+    target_state: ndarray = None
+    A: ndarray = None
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.target_state = self.backend.cast(self.target_state, dtype=self.backend.np.complex128)
+        self.A = self.backend.cast(self.A, dtype=self.backend.np.complex128)
+
+        
+    def __call__(self, circuit: Circuit):
+        result = super().__call__(circuit)
+        state = result.state()
+
+        if self.A is None or self.target_state is None:
+            raise_error(ValueError, "Both ``A`` and ``target_state`` must be provided.")  
+        
+        final_state = self.A @ state
+        normalized = final_state / self.backend.calculate_vector_norm(final_state)
+        overlap = self.backend.np.vdot(self.target_state, normalized)
+        cost = 1 - overlap.abs() ** 2
```

Comment:
```suggestion
        cost = infidelity(normalized, target, backend=self.backend)
```

While importing at top level `from qibo.quantum_info.metrics import infidelity`.